tests/pages/pageobjects.py

```diff
@@ -110,6 +108,59 @@ def fav_button_counter_text(self):
+                },
+                {
+                    "operationName": "besidaQuery",
+                    "variables": {},
+                    "query": "query besidaQuery {\n  besida {\n    cdn_url\n    desktop_v\n    mobile_v\n    is_besida_enabled\n    __typename\n  }\n}\n"
+                }
+            ])
+
+            headers = {
+                'Cookie': '; '.join([cookie['name'] + '=' + cookie['value'] for cookie in self.driver.get_cookies()]),
+                'x-requested-with': 'XMLHttpRequest',
+                'Content-Type': 'application/json'
+            }
+            response = requests.request("POST", url, headers=headers, data=payload)
+            is_in_fav = True if BasePage.good_name_text(operable_goods) in response.text else False
+            return is_in_fav ^ is_prev_heart_state
+
+        operable_goods = web_item_list[goods_index]
+        is_heart_picked_prev = is_heart_picked()
+        click_time = datetime.datetime.now().timestamp()
+        operable_goods.find_element(By.XPATH, './/span[@data-qaid="add_favorite"]').click()
+        while not is_heart_state_changed(is_heart_picked_prev):
+            if datetime.datetime.now().timestamp() - click_time > 1:
+                # done 1s for operation of adding/removing one goods to/from the favorite list
+                # if didnt raises assert exception and stops the test
+                assert False, 'Global problem about adding/removing goods to/from favorites.' \
```

Comment:
as a common convention, we try to avoid using asserts inside page objects, because that's usually is like an error - I mean it could be in one or two related tests, but in most other cases you just try to re-use that element and not testing it in every next test.
In other words - every test must have only one or a couple of asserts related to the tested logic of a particular test, all other issues and failures have to be exposed as errors.

Reply:
I suppose I have to raise error instead of assert, haven't me?

Reply:
assert changed to raising error, added doing screenshot before raising error